### ISSUE TITLE
PDF content blocked by Chrome

### ISSUE THREAD

**Issue by rgaudin** [kiwix/libkiwix#916]
When opening a PDF in a ZIM in the kiwix-serve viewer, on Chrome, there is an unrecoverable error message. This looks like a regression.

<img width="1087" alt="Screenshot 2023-03-15 at 14 18 57" src="https://user-images.githubusercontent.com/57929/225337621-43ffda0e-a0aa-45a1-b6ca-af715e4ff55f.png">

Go to https://dev.library.kiwix.org/viewer#lilote_fr_fo_2023-03/xay-va-%C3%A0-la-p%C3%AAche then click the _Lire l'histoire_ button.

**Comment by rgaudin** [comment]
Might be related to #604

**Comment by mgautierfr** [comment]
I have no problem with firefox, chrome and chromium.

**Comment by rgaudin** [comment]
@Popolechien reported it ; I can reproduce on latest Chrome

**Comment by rgaudin** [comment]
Sorry the link is incorrect ; you have to use the viewer. I need to open a ticket about that as well

**Comment by rgaudin** [comment]
https://dev.library.kiwix.org/viewer#lilote_fr_fo_2023-03/xay-va-%C3%A0-la-p%C3%AAche

**Comment by mgautierfr** [comment]
Indeded. Removing the sandbox attribut solve the problem.
This is probably introduced by https://github.com/kiwix/libkiwix/pull/906 (issue https://github.com/kiwix/kiwix-tools/issues/604)

**Comment by Jaifroid** [comment]
The code I give in that issue works around this specific problem with the pdf viewer in Chrome/Edge. It is not recognised by the browser as same origin, hence PDFs must be opened in a new window or tab.

**Comment by Jaifroid** [comment]
Personally I think this is a Chromium bug, because the in-browser PDF viewer should clearly be same-origin when loading a PDF from the same origin. Or maybe it's a feature because PDFs can have active content that can contact external servers (?).

I didn't find any other workaround in Kiwix JS than to make click on a PDF open a new window/tab. Having said that, the new window uses the same in-built PDF viewer, so it's not too big a deal. Epubs have to be downloaded anyway because no browser provides a custom viewer.

It's a balance between the security of the iframe (in terms of not leaking info out: a particular concern with Zimit archives) and (in)convenience... Otherwise, there's really nothing to stop a script in the iframe navigating elsewhere.

NB sandboxing the iframe doesn't stop a determined malicious attacker, but it stops accidental redirects, accidental (or well intentioned) attempts by scripts to break out of iframes, and accidental contacting of external sites for font files, images and (potentially) scripts.

**Comment by mgautierfr** [comment]
See https://github.com/whatwg/html/issues/3958 which give some information.

PDF viewer is implemented as a plugin in chrome and it is deactivated in sandboxed iframe.

**Comment by danielzgtg** [comment]
Please also add `allow-downloads`. Some people may have a Firefox setup where "Settings > General > Files and Applications > Applications > Portable Document Format (PDF)" is set to "Save File" instead of "Open in Firefox". I heard there was a recent proposal to use ZIMs for serving .apk files, and this would be necessary to support their use case.

The sandbox attribute needs to stay to stop the Wiktionary zim from breaking. What's wrong with pdf.js? It would only make the experience more consistent across platforms.

**Comment by veloman-yunkan** [comment]
I guess that a fix addressing #912 but not this ticket doesn't make much sense. On the other hand, I think that a fix to this issue should also automatically fix #912.

**Comment by Jaifroid** [comment]
This is what we went for in Kiwix JS (for the sandbox attribute of iframe, or can be served as part of CSP response header):

> `allow-same-origin allow-scripts allow-modals allow-forms allow-popups allow-downloads`

**Comment by Jaifroid** [comment]
@veloman-yunkan Is this sandbox attribute actually in force at library.kiwix.org yet? Because if you go to https://library.kiwix.org/viewer#wiktionary_en_all_nopic_2023-02 , you clearly see that a top-level navigation occurs and the iframe is destroyed..., and if I inspect the iframe just before the offending script runs that breaks out of it, I see what's in the screenshot below.

Now, if the sandbox isn't implemented at library.kiwix.org, it means that the issues with the PDF viewer are not directly down to #906, unless I'm missing something (or my browser cache is VERY persistent, despite clearing it)...

(Though I do think #906 will block PDFs in Chrome, because I already had to patch that in Kiwix JS.) Confused 😖...

![image](https://user-images.githubusercontent.com/4304337/225568099-0f69a150-f161-4da0-a643-08ff4ed28eab.png)

**Comment by Jaifroid** [comment]
And, btw, that `referrerpolicy` should probably be `none`, rather than `same-origin`.

**Comment by veloman-yunkan** [comment]
@Jaifroid https://library.kiwix.org runs the latest release of `kiwix-serve`. #906 has not been released yet. Its side effects are currently demonstrated at https://dev.library.kiwix.org (to which this ticket refers).

**Comment by Jaifroid** [comment]
For anyone coming to this longish thread at this point, the executive summary is as follows:

* There is a vulnerability in Kiwix Serve and Kiwix JS related to the use of iframes to display articles: scripts in these articles can accidentally (or on purpose) navigate to remote sites and leak user data to remote servers, or they can break out of the iframe and destroy the app's controls;
* A sandbox has therefore been added to the iframe to plug this vulnerability. This sandbox will not stop a purposefully malicious ZIM, but it will stop accidental navigation (as happens with recent Wiktionary), or accidental leakage (as happens when downloading remote fonts or, occasionally, images);
* However, adding this sandbox proves to have a side effect that in Chromium browsers, the built-in PDF viewer is untrusted (by design, because PDFs can have active content), and PDFs are therefore blocked from rendering in a sandboxed iframe;
* It is proposed instead that if a user clicks on a PDF link, the PDF will be opened in a new tab or window (not subject to sandboxing);
* This is not the behaviour the creator of the ZIM may have intended, so there is a balance to be struck between security and respect for the intentions of the ZIM creator / trusted web resource.

**Comment by mgautierfr** [comment]
> There is a vulnerability in Kiwix Serve and Kiwix JS related to the use of iframes to display articles: scripts in these articles can accidentally (or on purpose) navigate to remote sites and leak user data to remote servers, or they can break out of the iframe and destroy the app's controls;

No. The vulnerability is that we display a "falsely offline" version of website which can still phone home  and leak user data to remote servers. This is how web works and except by blocking all connections to server, it can always append.

Before iframe, kiwix-serve was simply displaying the content in the top frame. "falsely offline" websites were obviously able to phone home. And as we were inserting the top bar in the content of the page, it was even simpler for the website to break out the app's controls.
The iframe is the occasion to add some kind of protection against accidentally "falsely offline" website when it was impossible to do before. The change to iframe is not the source of the leak.
However, **we have exchange the website css breaking the css of the app controls against links with `target=_top` breaking the app controls**.

There is no more user data leak than before.
You can use the "no viewier/iframe" with the `/content/zim_name/path/to/article` to see that website can phone home all the time.

Blocking this requests was never a goal of kiwix-serve. It may change, but if it became a purpose, the solution is probably more in service worker (to control all connection going out of the displayed website) than in iframe.

> This is not the behaviour the creator of the ZIM may have intended, so there is a balance to be struck between security and respect for the intentions of the ZIM creator / trusted web resource.

There is a balance between adding a new security level and not inferring in the website content (in chrome browser)

---

Blocking any link with `target=_top` (with sandboxed iframe) may not be a good idea neither. 
A website may be internally composed of several iframes. On iframe may display a menu with links changing the top iframe content. Allowing `target=_top` link breaks the (viewer) ui, but at least the website is usable.
And this is difficulty patchable as other player than `kiwix-serve`/`kiwix-js` don't use iframe and so `target=_top` is totally valid.
It would be to the viewer itself to dynamically patch (or intercept) links with `target=_top` and replace them with `target=framename`.

However, if the `target=_top` is not really needed, we can consider that as a bug in the scrappers and they should remove it.

**Comment by Jaifroid** [comment]
@mgautierfr in https://github.com/kiwix/kiwix-tools/issues/604#issuecomment-1460035994 I suggested that instead of using the sandbox attribute on the iframe, Kiwix Serve can serve all content with a CSP sandbox response header. I agree with you that this would be conceptually more elegant. However, we would still have this issue with PDF content not being rendered in Chromium browsers in the iframe, and also the issue with external links being blocked. Like in Kiwix Android, these will still have to be intercepted and opened in an external window. Adding the sandbox attribute in the response header or adding it in the iframe are exactly equivalent in terms of functionality (I've tested this in the PWA).

**Comment by danielzgtg** [comment]
> No. The vulnerability

There are multiple versions of the bug. There is more information on Slack.

> website css breaking the css of the app controls

Shadow DOM was designed for this problem.

> a balance between adding a new security level and not inferring in the website content (in chrome browser)

And between fixing the website content (in Wiktionary)

> A website may be internally composed of several iframes

Have we seen how archive.org does it? archive.org edits the HTML of all webpages to insert its header with the controls. It rewrites links, and this is to ensure multimedia paths work and external links are intercepted. With that strategy, we can either use the CSP HTTP header like you suggested, or we can inject a `<meta>` tag. archive.org does not edit the data of multimedia such as images, so PDFs should be rendered properly if we switch to that strategy.

**Comment by Jaifroid** [comment]
@danielzgtg I used to use only a `<meta>` tag CSP in Kiwix JS PWA, but unfortunately you can't use the sandbox attribute in `<meta http-equiv...>`. See [Content-Security-Policy/sandbox](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Security-Policy/sandbox):

> This directive is not supported in the [`<meta>`](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/meta) element or by the [Content-Security-policy-Report-Only](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Security-Policy-Report-Only) header field.

And without `sandbox` we can't stop top-level navigation. The other disadvantage of CSPs via `<meta>` is that you have to alter the HTML to inject the tag, whereas adding a CSP response header or iframe sandbox does not require altering the HTML in any way.

Rewriting links isn't necessary if you [intercept the user's click](https://github.com/kiwix/kiwix-tools/issues/604#issuecomment-1470013026) on the iframe and inspect the target.

**Comment by mgautierfr** [comment]
> There are multiple versions of the bug. There is more information on Slack.

Please put it somewhere on github (here or on https://github.com/kiwix/kiwix-tools/issues/604) has we need trace of the discussion. Slack is not public.

> Shadow DOM was designed for this problem.

Give me more

> Have we seen how archive.org does it? archive.org edits the HTML of all webpages to insert its header with the controls. It rewrites links, and this is to ensure multimedia paths work and external links are intercepted. With that strategy, we can either use the CSP HTTP header like you suggested, or we can inject a <meta> tag. archive.org does not edit the data of multimedia such as images, so PDFs should be rendered properly if we switch to that strategy.

We were modifying the content in kiwix-serve but we changed that especially to not modifying it. The idea is to interfere the less possible with the content in the zim file as this content should be "just working". (If not, fix the scrapper).

**Comment by mgautierfr** [comment]
> There is a good shorter-than-1-page demonstration at [javascript.info/shadow-dom#encapsulation](https://javascript.info/shadow-dom#encapsulation) .

I new world opens to me. Thanks for the link. It would have been nice to know about that few years ago

**Comment by Jaifroid** [comment]
On Shadow DOM, remember we need to support a wide range of browsers on possibly quite old devices, given where Kiwix Serve needs to be deployed, and although shadow DOM is now quite widely supported, that wasn't the case a few years ago, and it's still the case that support is patchy on mobile device browsers, for example. For maximum compatibility, I think we're still stuck with iframes. See https://caniuse.com/shadowdomv1 .

**Comment by danielzgtg** [comment]
> wide range of browsers on possibly quite old devices, [...] on mobile device browsers

Old mobile devices aren't a problem. Chrome is evergreen/autoupdatingo for Android 5+. My Firefox Nightly is Android 5+. It's evergreen too, and the only non-evergreen Firefox is Firefox ESR, but by its version number, it supports Shadow DOM too. Opera Mini is not a concern because it requires internet access to Opera servers. The only concern might be Android Browser / Android Webview. That is tied to the OS and used when the phone does not have Google Play Services.

> we need to support a wide range of browsers

IE11 is the only problematic browser left. If that's the case we can include a shadow DOM polyfill. A quick Google search led me to https://github.com/webcomponents/polyfills/tree/master/packages/shadydom and https://github.com/tuespetre/shadow-dom .

@Jaifroid Can you commit a `.browserslistrc` file somewhere? I tried looking for it in kiwix-js, kiwix-js-windows, libkiwix, and kiwix-tools. kiwix-js only describes it in informal text and the versions written there seem to be higher than what you mean by "quite old devices". Both developers and automated tools can understand the https://github.com/browserslist/browserslist standard. Having that file would bring everyone onto the same page about what browsers are need to be supported.

**Comment by rgaudin** [comment]
> Old mobile devices aren't a problem. Chrome is evergreen/autoupdatingo for Android 5+

This assumes Internet access which lack is precisely why we deploy Kiwix

**Comment by Jaifroid** [comment]
Coming back to the main purpose of this issue, I think I have a solution that allows us to sandbox ZIM articles and does not interfere with loading PDFs into the iframe. It is the one I outlined above, i.e. that we serve articles from the ZIM with a CSP sandbox response header, but not PDFs, and we remove the sandbox attribute from the iframe. I have a proof-of-concept here using the Kiwix JS browser extension's test PWA:

* https://kiwix.github.io/kiwix-js/www/index.html

Please ensure this is showing version 3.7.3 (if it's not, wait around 10 seconds for a notification that 3.7.3 is ready to load). This version correctly blocks the attempt at top-level navigation by the February English Wiktionary ZIM, but allows PDFs to load in the iframe. These are the settings used (some incidental things are specific to Kiwix JS):

### Response headers

```
// Set Content-Security-Policy to sandbox the content (prevent XSS attacks from malicious ZIMs)
headers.set('Content-Security-Policy', "default-src 'self' data: blob: about: chrome-extension: https://moz-extension.kiwix.org 
    https://kiwix.github.io 'unsafe-inline' 'unsafe-eval'; sandbox allow-scripts allow-same-origin allow-modals allow-popups 
    allow-forms allow-downloads;");
headers.set('Referrer-Policy', 'no-referrer');
```

### Iframe

`<iframe id="articleContent" class="articleIFrame" src="article.html" referrerpolicy="no-referrer"></iframe>`

### Meta http-equiv CSP (in the HTML document that contains the iframe)

```
<meta http-equiv="Content-Security-Policy" content="default-src 'self' data: https://download.kiwix.org
    https://master.download.kiwix.org https://moz-extension.kiwix.org https://kiwix.github.io 'unsafe-inline' 'unsafe-eval'; 
    frame-src 'self' moz-extension: chrome-extension:; object-src 'none';">
```

I hope this is helpful in resolving this long-winded issue and also #604 / #906.

**Comment by kelson42** [comment]
Thank you @Jaifroid for trying to wrap-up. I have a hard time to get a ckear picture myself. But is seems to me to be the best path. Anybody has a better alternative in mind?

**Comment by Jaifroid** [comment]
Since he did #906, maybe @veloman-yunkan could kindly try a dev implementation of [the above](https://github.com/kiwix/libkiwix/issues/916#issuecomment-1483102716) for Kiwix Serve, removing the Kiwix JS specific code, and setting the CSP/sandbox response header in the server code (not via a Service Worker). My test implementation (done with Kiwix JS) doesn't test the case of Zimit archives, because the Kiwix JS browser extension doesn't run those yet, so it would be important to test those thoroughly. The question of opening external links in a new tab/window may also need to be handled if Kiwix Serve doesn't already have a way of doing this, but it's not difficult and there is some suggested code in https://github.com/kiwix/kiwix-tools/issues/604#issuecomment-1470013026 (NB the part of that code dealing with PDFs should be removed).

**Comment by veloman-yunkan** [comment]
> I don't understand the mechanics of your solution

@Jaifroid By above I meant that I don't see why the described setup should enable the chrome PDF extension (that is not same-origin) being loaded in an iframe that only allows same-origin content.

**Comment by Jaifroid** [comment]
@veloman-yunkan The test implementation was erased yesterday by a push to master of a completed PR. However, I've just restored it, so if you visit that address again, and wait 10s or so, it should offer you v3.7.3 (see screenshot). To load, exit the browser and then re-visit that page. It should then show 3.7.3 and allow you to test.

Regarding setting the headers in Kiwix Serve, how are you doing this? And are you setting response headers only for HTML? I suspect you need to set different headers (or none) for a PDF, but I don't know that for sure without testing empirically.

**Comment by Jaifroid** [comment]
Here is top-level navigation (English Wiktionary) being blocked by 3.7.3, and below that is a PDF (Greek Gutenberg) showing in the iframe in Edge (Chromium) also on 3.7.3:

![image](https://user-images.githubusercontent.com/4304337/228007070-4e6885db-fbda-4881-a6a4-1bdbea3a3bf0.png)

![image](https://user-images.githubusercontent.com/4304337/228007216-7625761f-1b28-40a2-a727-fac4cca5ade0.png)

**Comment by veloman-yunkan** [comment]
> @veloman-yunkan The test implementation was erased yesterday by a push to master of a completed PR. However, I've just restored it, so if you visit that address again, and wait 10s or so, it should offer you v3.7.3 (~see screenshot~). To load, exit the browser and then re-visit that page. It should then show 3.7.3 and allow you to test.

@Jaifroid Thanks. The 3.7.3 PWA doesn't load PDFs in the iframe either (when using chromium).
 
> Regarding setting the headers in Kiwix Serve, how are you doing this? And are you setting response headers only for HTML? I suspect you need to set different headers (or none) for a PDF, but I don't know that for sure without testing empirically.

Let's dig into this after you confirm that the PWA works for you using this ZIM file: https://master.download.kiwix.org/zim/.hidden/dev/lilote_fr_test_2023-01.zim

**Comment by Jaifroid** [comment]
Working fine with that ZIM. Are you sure the app is showing 3.7.3 in top-left corner?

![image](https://user-images.githubusercontent.com/4304337/228009004-b7a4eef7-3ab7-4454-8bd5-04d15afc1f11.png)

**Comment by Jaifroid** [comment]
Also working in Chrome 110 (previous screenshots were Edge Chromium 111):

![image](https://user-images.githubusercontent.com/4304337/228013549-039890bb-eb95-4197-aea2-fbe83a14d5e5.png)

**Comment by veloman-yunkan** [comment]
Doesn't work with Chromium Version 90.0.4430.212 (installed on Ubuntu 20.04 using [these instructions](https://askubuntu.com/a/1206153))

![Screenshot from 2023-03-27 22-26-04](https://user-images.githubusercontent.com/11933502/228033123-afeda931-6381-4950-a27b-5026b8d12ca2.png)

**Comment by Jaifroid** [comment]
OK, I'll see if I can reproduce and debug with Chrome 90 on Windows.

**Comment by kelson42** [comment]
Thx for investogating this in details, but pretty possible that a bug in Chrome has been fixed between versions 90 and 110.

**Comment by Jaifroid** [comment]
Yes, I've just checked on Browser Stack, and I confirm that the bug is present in Chrome 90, but it was fixed in Chrome 91 (pictured below) and later. Chrome 90 doesn't recognize the PDF loaded as being of origin 'self'.

![image](https://user-images.githubusercontent.com/4304337/228217686-e505f70f-c932-4d93-a471-549f0e80ac26.png)

**Comment by Jaifroid** [comment]
Yes, it's pretty recent, but it's actually quite hard to stay on Chrome 90, because Google updates it almost as soon as you install it. Of course that wouldn't be the case with Chromium on Linux.

I haven't spent a long time trying to find out if there's a workaround for Chrome 90, but I did try specifying the site (https://kiwix.github.io) as an allowed frame-src and even tried deleting the mtea http-equiv CSP, but the bug persists. The error in DevTools just confirms the block.

![image](https://user-images.githubusercontent.com/4304337/228221488-b4d08a98-64bb-49dd-a554-9cc270c5e084.png)

------------------------

Bottom line, unless someone has a specific patch for Chrome <=90's behaviour, it seems to me we have these choices:

1. Revert sandboxing added in #906, leaving the Kiwix Serve iframe vulnerable to top-level navigation like in the February Wiktionary (which started all this!) and XSS;
2. Adopt the fix we're testing here (needs further careful testing with different ZIMs), but advise users of Chrome <= 90 that they will need to upgrade if they want to view PDFs;
3. Add the workaround which opens any PDF in a separate tab or window (sample code provided here: https://github.com/kiwix/kiwix-tools/issues/604#issuecomment-1470013026). NB workaround is only for PDFs opened with a user gesture.

In Kiwix JS we decided to adopt 3 (actually version 2 + 3), as it is the most universal solution, even if it doesn't mimic precisely what the original web site developers may have intended. Users are forced to open EPUBs in a separate app, so opening PDFs in a separate tab is not so much of a sacrifice IMHO.

**Comment by veloman-yunkan** [comment]
> Explanation: I realized that the Service Worker was setting the sandbox response headers for PDF files as well as HTML files, so I added the code below exempting PDFs. Something similar could be done in the back-end code of Kiwix Serve.

@Jaifroid That was already the case in my local modification to `kiwix-serve`. I also added the kiwix-serve host address (`http://localhost:8080`) to `default-src` in both `Content-Security-Policy` header and the meta http-equiv CSP, but it didn't help. Can the fact that the content is served over http rather than https be a problem?

**Comment by veloman-yunkan** [comment]
My changes (without the temporary hack of hardcoding `http://localhost:8080` in `default-src`) can be seen in #924

**Comment by Jaifroid** [comment]
> That was already the case in my local modification to `kiwix-serve`. I also added the kiwix-serve host address (`http://localhost:8080`) to `default-src` in both `Content-Security-Policy` header and the meta http-equiv CSP, but it didn't help. Can the fact that the content is served over http rather than https be a problem?

@veloman-yunkan Localhost is always http but it is considered secure (for testing purposes) by browsers, though I guess we can't rule out a bug here in different browsers. I see you removed the sandbox on the iframe, so it is really puzzling we're getting different results for different implementations. Are all caches fully cleared when you're testing? Might just be worth checking in dev tools that there is no sandbox attribute still set on the iframe... Otherwise, I think we'd need a test implementation to debug what's going on.

**Comment by veloman-yunkan** [comment]
> I see you removed the sandbox on the iframe, so it is really puzzling we're getting different results for different implementations. Are all caches fully cleared when you're testing? Might just be worth checking in dev tools that there is no sandbox attribute still set on the iframe...

Just checked in dev tools that there is no `sandbox` attribute set on the iframe.
  
> Localhost is always http but it is considered secure (for testing purposes) by browsers

That's nice but we need `kiwix-serve` to also work in a LAN over http. So does your answer imply that it may be a problem?

**Comment by Jaifroid** [comment]
> That's nice but we need `kiwix-serve` to also work in a LAN over http. So does your answer imply that it may be a problem?

I don't think we're requiring secure origins as part of the sandbox, but we'd need to check the documentation to see if that is enforced or not. Or else test it empirically. However, I can definitively say that v3.7.5 of Kiwix JS works exactly the same whether served from localhost or served from https://kiwix.github.io, so I don't think this has anything to do with the reason we are seeing differences between Kiwix JS and Kiwix Serve despite both trying to serve the same headers.

One unrelated issue with serving ZIM content over LAN using http is that Zimit archives, which depend on a Service Worker (requiring a secure origin) won't work. Zimit archives **do** work over http://localhost, however, as this is considered secure.

**Comment by danielzgtg** [comment]
> It is the one I outlined above, i.e. that we serve articles from the ZIM with a CSP sandbox response header, but not PDFs

Thank you for this alternative fix. I confirmed that my Wiktionary use case still works on 3.7.5.

> Only applicable to Desktop users with Online access

Only applicable if we don't accept my suggestion to just bundle the Chrome/Firefox installers for Windows/macOS/Linux/Android

> security hole for PDFs, which would no longer be sandboxed

`Content-Type` should already prevent any PDF from being run as HTML/JS. And unless we bundle pdf.js, we can only assume that Chrome sandboxs PDFs to not run Javascript and to not make network requests

> I can't see anything in https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Security-Policy/sandbox that presumes a secure origin.

See https://developer.mozilla.org/en-US/docs/Web/Security/Secure_Contexts

It is possible to add a custom domain to `/etc/hosts` if you don't have two computers to test with

**Comment by rgaudin** [comment]
> Only applicable if we don't accept my suggestion to just bundle the Chrome/Firefox installers for Windows/macOS/Linux/Android

This is hardly realistic to expect users to download and install and use a browser ; at least in all the deployment scenarios I've seen. Not to mention the potential system version conflict that might exist. I doubt latest Chrome can be installed on Android 5.1

Should it be done, it should be external to kiwix-serve anyway. We do offer some apps to download. It would be instructive to test with a browser though.

**Comment by rgaudin** [comment]
It just came to my attention that we have a somewhat similar issue with EPUB with **firefox**

First, verify the kiwix-serve 3.4.0 behavior by going to [library](https://library.kiwix.org/viewer#gutenberg_mul_all_2023-03/Home.html). Click the HTML button of any book, it works. Go back and click an EPUB link, it should either open in your in-browser epub extension if you have one or trigger a download.

Now, using kiwix-serve nightly, go to [dev-library](https://dev.library.kiwix.org/viewer#gutenberg_en_all_2023-03/Home.html)

Reproduce those steps and you'll notice that nothing happens when you click on the EPUB link. This is even worst UX than the PDF behavior because there is no feedback and the user would only assume that the ZIM is broken.

You can right-click and open in a new tab, and it would work as expected.

As you'd imagine, the console mentions the culprit:

```
Download of “https://dev.library.kiwix.org/content/gutenberg_en_all_2023-03/Romeo%20and%20Juliet.1513.epub” was blocked because the triggering iframe has the sandbox flag set.
```

Same behavior with Chrome:

```
Download is disallowed. The frame initiating or instantiating the download is sandboxed, but the flag ‘allow-downloads’ is not set. See https://www.chromestatus.com/feature/5706745674465280 for more details.
```

**Comment by Jaifroid** [comment]
The code I suggested explicitly includes "allow-downloads", but it's currently only in a PR #924, and not yet in the dev version. It would be useful if @veloman-yunkan could merge that code so we can thoroughly test and debug it in the dev server for issues such as this.

**Comment by Jaifroid** [comment]
I'm re-opening because it's not fixed. @veloman-yunkan, the main difference I can see, at first sight, between the implementation at https://kiwix.github.io/kiwix-js/ and the latest Kiwix Serve at https://dev.library.kiwix.org/ is that in the latter case the sandbox is applied to the top-level document, i.e. to the `viewer` document (see screenshot). In the Kiwix JS version, I removed the top-level sandbox, so that it could be applied on a per-document basis to the content in the iframe. Setting it at top level probably means it can't be turned off when a PDF is being loaded in the iframe. Could you try removing the sandbox on `viewer` (but leave it for any content other than PDFs loaded from the ZIM)?

![image](https://user-images.githubusercontent.com/4304337/230782053-02cac1c3-7401-46ae-bbb5-3ddb383e2647.png)

**Comment by rgaudin** [comment]
@Jaifroid please elaborate on how it's not fixed. @kelson42 and I tested the use cases in the ticket and both were resolved

**Comment by Jaifroid** [comment]
@rgaudin Assuming the code in https://dev.library.kiwix.org, after the merging of #924, is the same code you have been testing, then as shown in the screenshot above, PDFs do not load in the iframe in Chromium browsers, whereas in the test case at https://kiwix.github.io/kiwix-js/, they do. @veloman-yunkan had also [said that it was not working](https://github.com/kiwix/libkiwix/pull/924#issue-1645849218). I believe the reason may be because the HTML document that contains the iframe is sandboxed in the dev version of Kiwix Serve, whereas it is not sandboxed in the Kiwix JS test case. The sandbox response header should only be served for content from the ZIM that is injected into the iframe (with PDFs excluded).

**Comment by kelson42** [comment]
@Jaifroid What are the reproduction steps? unclear for me as well.

**Comment by rgaudin** [comment]
Indeed, we tested both issues with Firefox while the first one is Chrome-only.

**Comment by veloman-yunkan** [comment]
@Jaifroid Thanks for testing and identifying the difference. In my "fix" (#924) I incorrectly apply the "Content-Security-Policy" header to *all* non-PDF responses. I will now add that header only to content intended to be displayed in the viewer (served via the `/content` endpoint).
